tasks/check-and-prepare-role-variables.yml

```diff
@@ -9,20 +9,23 @@
       run_once: yes
   when: ha_cluster_cluster_present
 
-- name: Extract all node names
+- name: Discover cluster node names
   set_fact:
     __ha_cluster_node_name: "{{ \
-        hostvars[item].ha_cluster.node_name | default(item) \
+        hostvars[inventory_hostname].ha_cluster.node_name \
+        | default(inventory_hostname) \
       }}"
-  loop: "{{ ansible_play_hosts }}"
-  run_once: yes
+
+- name: Extract cluster node names
+  set_fact:
+    __ha_cluster_all_node_names: "{{ hostvars[item].__ha_cluster_node_name }}"
+  loop: "{{ ansible_play_hosts_all }}"
```

Comment:
I'm not sure what you are trying to do here but it doesn't look right to me - `__ha_cluster_all_node_names` will not be a list, it will be a string - can you explain what you are trying to do here?

Reply:
There are three tasks which work together to construct the list of node names.

`Discover cluster node names` - figures out a node name for each host, that is either an inventory variable or `inventory_hostname`. This is actually used in tests to replace `/cache/centos8.qcow2` with node names sensible from the HA cluster point of view. It also ensures that `__ha_cluster_node_name` is always defined.
```
ok: [rh83-node2] => {"ansible_facts": {"__ha_cluster_node_name": "rh83-node2"}, "changed": false}
ok: [rh83-node3] => {"ansible_facts": {"__ha_cluster_node_name": "rh83-node3"}, "changed": false}
```

`Extract cluster node names` - by looping over `ansible_play_hosts_all`, I get all node names into one 'variable'. The fact itself is not important here at all and it gets overwritten by the following task. What matters is the registered variable `__ha_cluster_node_name_result`, which is processed by the following task. Yes, `__ha_cluster_all_node_names` will be a string, but it will be a string for each node.
```
ok: [rh83-node2] => (item=rh83-node2) => {"ansible_facts": {"__ha_cluster_all_node_names": "rh83-node2"}, "ansible_loop_var": "item", "changed": false, "item": "rh83-node2"}
ok: [rh83-node2] => (item=rh83-node3) => {"ansible_facts": {"__ha_cluster_all_node_names": "rh83-node3"}, "ansible_loop_var": "item", "changed": false, "item": "rh83-node3"}
ok: [rh83-node3] => (item=rh83-node2) => {"ansible_facts": {"__ha_cluster_all_node_names": "rh83-node2"}, "ansible_loop_var": "item", "changed": false, "item": "rh83-node2"}
ok: [rh83-node3] => (item=rh83-node3) => {"ansible_facts": {"__ha_cluster_all_node_names": "rh83-node3"}, "ansible_loop_var": "item", "changed": false, "item": "rh83-node3"}
```
`Collect cluster node names` - collects node names from the registered result of `Extract cluster node names` to create a list of strings - node names. Since this works based on `__ha_cluster_node_name`, which is always defined by the first task, `__ha_cluster_all_node_names` is always defined as well and it contains all nodes.
```
ok: [rh83-node2] => {"ansible_facts": {"__ha_cluster_all_node_names": ["rh83-node2", "rh83-node3"]}, "changed": false}
ok: [rh83-node3] => {"ansible_facts": {"__ha_cluster_all_node_names": ["rh83-node2", "rh83-node3"]}, "changed": false}
```

Perhaps I could have named the fact in `Extract cluster node names` differently to make it more clear.

It seems quite complicated to me to have these three tasks to create a single list of node names. It's the best I was able to figure out, but my ansible skills are still not the best. Let me know if you know a simpler way to achieve this.

Reply:
ok - so you need to have some variable e.g. `__ha_cluster_all_node_names` in the hostvars of each host.  This variable contains the list of all of the ha_cluster node names for all of the hosts in the cluster.

I believe `hostvars[inventory_hostname].ha_cluster.node_name` is exactly the same as `ha_cluster.node_name`.

Let me see what I can come up with.

Reply:
you only need 1 task
```yaml
- name: Collect cluster node names
  set_fact:
    __ha_cluster_all_node_names: "{{
      ansible_play_hosts_all | map('extract', hostvars, ['ha_cluster', 'node_name']) |
      map('default', inventory_hostname) | list
    }}"
```
See https://docs.ansible.com/ansible/2.8/user_guide/playbooks_filters.html#extracting-values-from-containers
Here is some sample code https://gist.github.com/richm/0f6dd759a461705945eb5dd0c5287815
I run it like this: `ANSIBLE_STDOUT_CALLBACK=debug ansible-playbook -vv ha_cluster_test.yml`
the output looks like this:
```
TASK [debug] *******************************************************************
task path: /home/rmeggins/ansible_sandbox/ha_cluster_test.yml:19
ok: [localhost] => {}

MSG:

ansible_play_hosts_all ['localhost', 'node1', 'node3', 'node2'] hostvars ['localhost', 'node1-ha', 'node3-ha', 'localhost']
```
I'm assuming in a "real" environment there won't be a `localhost` nor duplicate hosts - if duplication is a problem, there's probably something else going on, but at any rate you can use the `unique` filter if you think it could be a problem

Reply:
Thank you very much, this works perfectly! `Discover cluster node names` task is actually needed as well, because `__ha_cluster_node_name` fact it sets is used in the role. But we got rid of the ugly `Extract cluster node names` task.

`unique` is not used on purpose. If there is a duplicity in node names, it will get discovered and reported by `pcs` as an error. That is a correct behavior, node names in a cluster must be unique. If they are not, then there is either a bug in the role or the user provided duplicate names. Both cases should be reported as an error.